### PR TITLE
BUG: MultiIndex mangling during parsing (#18062)

### LINKED ISSUE #18062: Reading a CSV with duplicated MultiRow columns
#### Problem description
I have the following .csv file:

A | A | B
-- | -- | --
one | one | two
0 | 34 | 0.1

I want to use the first two rows to create MultiIndex columns. However, the first and second columns are duplicated and pandas is handling that in a weird way.

#### Code Sample, a copy-pastable example if possible

```python
print(pd.read_csv('test.csv',header=[0,1]).columns)
>> Index([('A', 'one'), '('A', 'one').1', ('B', 'two')], dtype='object')
```
This output is a total mess, the first column and last are tuples while the second column is a string. Also, this is not a MultiIndex, it's an Index. Raising an error or having the option to delete the duplicate columns would be a better solution.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-97-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.21.0
pytest: 2.9.2
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.24.1
numpy: 1.13.3
scipy: 0.18.1
pyarrow: None
xarray: None
IPython: 5.1.0
sphinx: 1.4.6
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.1.0
tables: 3.2.3.1
numexpr: 2.6.1
feather: None
matplotlib: 1.5.3
openpyxl: 2.3.2
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.3
lxml: 3.6.4
bs4: 4.5.1
html5lib: None
sqlalchemy: 1.0.13
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #18062 
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -1106,6 +1106,11 @@ def _is_index_col(col):
     return col is not None and col is not False
 
 
+def _is_potential_multi_index(columns):
```
Comment (gfyoung) [inline pandas/io/parsers.py]: Docstring

File: pandas/tests/io/parser/header.py [hunk 2]
```diff
@@ -290,3 +290,13 @@ def test_singleton_header(self):
...
+
+    def test_mangles_multi_index(self):
+        # See GH 18062
+        data = """A,A,A,B\none,one,one,two\n0,40, 34,0.1"""
+        df = self.read_csv(StringIO(data), header=[0, 1])
+        expected = DataFrame([[0, 40, 34, 0.1]],
+                             columns=MultiIndex.from_tuples(
+                                 [('A', 'one'), ('A', 'one.1'),
+                                  ('A', 'one.2'), ('B', 'two')]))
+        tm.assert_frame_equal(df, expected)
```
Comment (gfyoung) [inline pandas/tests/io/parser/header.py]: Add a test for the following case:

```
data = """A,A,A,B\none,one,one.1,two\n0,40,34,0.1"""
```

File: doc/source/whatsnew/v0.22.0.txt [hunk 3]
```diff
@@ -89,6 +89,7 @@ Bug Fixes
 
 - Bug in ``pd.read_msgpack()`` with a non existent file is passed in Python 2 (:issue:`15296`)
 - Bug in ``DataFrame.groupby`` where key as tuple in a ``MultiIndex`` were interpreted as a list of keys (:issue:`17979`)
+- Bug in ``pd.read_csv`` where a multi-index with duplicate columns was not being mangled appropriately (:issue: `18062`)
```
Comment (jschendel) [inline doc/source/whatsnew/v0.22.0.txt]: Can you make the following changes: 
  - <code>\`\`pd.read_csv\`\`</code> -> <code>:func:\`read_csv\`</code>
  - "multi-index" -> <code>\`\`MultiIndex\`\`</code>
  - Remove the space between `:issue:` and the issue number

File: pandas/tests/io/parser/header.py [hunk 4]
```diff
@@ -290,3 +290,21 @@ def test_singleton_header(self):
...
         expected = DataFrame({"a": [0, 1], "b": [1, 2], "c": [2, 3]})
         tm.assert_frame_equal(df, expected)
+
+    def test_mangles_multi_index(self):
+        # See GH 18062
+        data = """A,A,A,B\none,one,one,two\n0,40, 34,0.1"""
+        df = self.read_csv(StringIO(data), header=[0, 1])
+        expected = DataFrame([[0, 40, 34, 0.1]],
+                             columns=MultiIndex.from_tuples(
+                                 [('A', 'one'), ('A', 'one.1'),
```
Comment (jreback) [inline pandas/tests/io/parser/header.py]: why is this not:

```
In [5]: MultiIndex.from_tuples([('A', np.nan), ('A', 'one'),('A', 'one'), ('B', 'two')]).values
Out[5]: array([('A', nan), ('A', 'one'), ('A', 'one'), ('B', 'two')], dtype=object)
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/parser/header.py (hunk 4 above)
Comment (WillAyd) [reply pandas/tests/io/parser/header.py]: Hey Jeff - I'm not sure I follow. Why would you expect there to be a NaN with those values? Or are you asking to add another test case that includes NaN values?

Comment (jreback) [reply pandas/tests/io/parser/header.py]: well the first value is missing, yes?

Comment (WillAyd) [reply pandas/tests/io/parser/header.py]: Shouldn't be. Assuming

    data = """A,A,A,B\none,one,one.1,two\n0,40, 34,0.1"""

Then the first row has 3 A's and a B. The second row has 2 one's, 1 one.1 and a two in the data, so both rows essentially have 4 columns and no missing data

Comment (jreback) [reply pandas/tests/io/parser/header.py]: hahha, was reading '\none' as none. ok then.

File: pandas/tests/io/parser/header.py [hunk 5]
```diff
@@ -290,3 +290,21 @@ def test_singleton_header(self):
...
+                                  ('A', 'one.2'), ('B', 'two')]))
+        tm.assert_frame_equal(df, expected)
+
+        data = """A,A,A,B\none,one,one.1,two\n0,40, 34,0.1"""
+        df = self.read_csv(StringIO(data), header=[0, 1])
+        expected = DataFrame([[0, 40, 34, 0.1]],
+                             columns=MultiIndex.from_tuples(
+                                 [('A', 'one'), ('A', 'one.1'),
+                                  ('A', 'one.1.1'), ('B', 'two')]))
+        tm.assert_frame_equal(df, expected)
```
Comment (gfyoung) [inline pandas/tests/io/parser/header.py]: Very nice!  Looking at your code, can we add one more test case here:

~~~python
data = """A,A,A,B\none,one,one.1,two,two\n0,40, 34,0.1,0.1"""
~~~

**Review by gfyoung — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@gfyoung merge if you are ok with this. lgtm.

**Comment by gfyoung** [comment]
Thanks @WillAyd !

**Pull request merged**